src/coreclr/jit/optimizer.cpp

```diff
@@ -4767,18 +4767,41 @@ bool Compiler::optIfConvert(BasicBlock* block)
     }
 #endif
 
-    // Duplicate the input of the assign.
+    // Duplicate the destination of the assign.
     // This will be used as the false result of the select node.
-    GenTree* destination = gtCloneExpr(asgNode->AsOp()->gtOp1);
-    destination->gtFlags &= GTF_EMPTY;
+    assert(asgNode->AsOp()->gtOp1->IsLocal());
+    GenTreeLclVarCommon* destination = asgNode->AsOp()->gtOp1->AsLclVarCommon();
+    GenTree*             falseInput  = gtCloneExpr(destination);
+    falseInput->gtFlags &= GTF_EMPTY;
+
+    // Create a new SSA entry for the false result.
+    {
+        // Get the SSA num of the destination.
+        unsigned lclNum            = destination->GetLclNum();
+        unsigned destinationSsaNum = destination->GetSsaNum();
+        assert(destinationSsaNum != SsaConfig::RESERVED_SSA_NUM);
+        LclSsaVarDsc* destinationSsaDef = lvaGetDesc(lclNum)->GetPerSsaData(destinationSsaNum);
+
+        // Create a new SSA num.
+        unsigned newSsaNum = lvaGetDesc(lclNum)->lvPerSsaData.AllocSsaNum(getAllocator(CMK_SSA));
+        assert(newSsaNum != SsaConfig::RESERVED_SSA_NUM);
+        LclSsaVarDsc* newSsaDef = lvaGetDesc(lclNum)->GetPerSsaData(newSsaNum);
+
+        // Copy across the SSA data.
+        newSsaDef->SetBlock(destinationSsaDef->GetBlock());
+        newSsaDef->SetAssignment(destinationSsaDef->GetAssignment());
+        newSsaDef->m_vnPair = destinationSsaDef->m_vnPair;
+        falseInput->AsLclVarCommon()->SetSsaNum(newSsaNum);
+
+        fgValueNumberTree(falseInput);
```

Comment:
It would be better to extract the `GT_LCL_VAR` case from `fgValueNumberTree` and call it directly here. `fgValueNumberTree` itself has other invariants this call is not ensuring (e.g. VN sets `compCurBB` and `compCurTree` around the call).